### ISSUE TITLE
Odata query substringof

### ISSUE THREAD

**Issue by hgc10120** [ghiscoding/Angular-Slickgrid#263]
Hi guys,

I'm currently using SlickGrid to display data retrieved from my server via the backend api service(Odata). Everything works accordingly, until I try to filter on a string column and "substringof" is used as a filter param. I'm using version 7.1.0 of Microsoft.AspNetCore.OData on the server side, and on this version, the query param of "substringof" is no longer in use. As a result, there is a discrepancy between what the client side(slickgrid) is sending when filtering string columns and what the server side(Microsoft.AspNetCore.OData) understands. Microsoft.AspNetCore.OData uses "contains" as opposed to "substringof" to query string columns. Is that an option that will be updated in the slick grid framework to allow to send filter queries of string columns with "contains", as opposed to "substringof".

Thanks.

**Comment by ghiscoding** [comment]
Hmm I wrote the code for OData version 2 I think, did that change with newer version of OData? If that is the case, we might need a new flag for the version. I really thought there was no differences in querying with all OData versions. oh and the code to build the OData query string is all done internally in the `GridOdataService` and `ODataQueryBuilder`, the `substringof` is on this [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/services/grid-odata.service.ts#L343). I'd be happy to have a PR (Pull Request) for a fix, I'm assuming adding a new flag to know the OData version is the best way to go.

Side note, when I created this OData Service we were using OData v2 with .Net 4.7, but we switched to GraphQL with .Net Core, so we are not using the OData anymore from our end.

**Comment by hgc10120** [comment]
It might have changed with the newer version. I've found a unsustainable way to go around it; i.e manually editing all queries that includes the string of 'substringof', but it isn't very scalable logic. I'll check out GraphQL as well. Do you mind letting me know if/when i can expect this issue to be resolved in the current version of OData used by Slickgrid?
Thanks.

**Comment by ghiscoding** [comment]
Well for that I really need to know if `substringof` really got replaced by `contains` and if so in what version of OData did that happen? I can't just replaced on the spot, that could potentially break others code if they use the older version of OData. Can you also provide a query string sample of your issue, what the grid gives you now and what you're expecting

**Comment by hgc10120** [comment]
Sorry about not using the template. It's my first time submitting an issue with github so i'm still learning all of it. After doing some research, it seems that contains replaced 'substringof' in V4 of Odata. When i send this query to the server  "$top=10000&$filter=(substringof('florida', Area)", I get an error message suggesting that 'substringof' is an unknown function. When I use contains instead, everything works accordingly.

**Comment by ghiscoding** [comment]
So the position of the operator is the same? I mean just replacing to `contains` and leaving the same position for the 2 arguments of it, is that correct? 

By the way, GitHub uses markdown (more info [here](https://guides.github.com/features/mastering-markdown/)) and you can use the back tick ` ` ` to show code or variable (that will show in grey in the post and is easier to read), or if you want multi-line code, you use 3 back ticks with the language for example
```ts
   ```ts
      someFunction() {
      }
 '  ``` // don't forget to close it with 3 back ticks again
```

**Comment by hgc10120** [comment]
Thanks for the tips. I'll be sure to check those out. The position of the 2 arguments are swapped. For instance, for this query to return what i need it to with contains, it would need to be written : `$top=10000&$filter=(contains(Area, 'florida')` . It would be relatively easy to replace the substring of part of the string with contains, if it didn't require that the arguments be swapped for the function to work correctly.
Thanks.

**Comment by ghiscoding** [comment]
@hgc10120 
I've create a PR #264 to address the issue, can you double-check the fix before I merge it, the info is all in there. I added a radio box in the Example 5 to switch between version 2 and 4, it will be part of the demo.

**Comment by hgc10120** [comment]
Just double checked everything and all of the info is there. Thank you for getting on this so quickly.

**Comment by ghiscoding** [comment]
Thanks for the double check, I'll be releasing that next possibly since I'm still working on a new Filter by Range feature, I should be done next week.

Cheers ⭐️

**Comment by ghiscoding** [comment]
@hgc10120 
So I found out that `contains` is not the only difference in OData 4 queries when working on the new date range feature, it seems that dates format changed as well, from `StartDate ge DateTime'2014-06-23T03:30:00.000Z'` to `StartDate ge 2014-06-23T03:30:00.000Z` as can be seen in this [SO](https://stackoverflow.com/a/31277779/1212166)

Can you tell me if this new query is valid? I don't have access to an OData server anymore but I would still like this to be working out of the box.

The new query for OData v4 is
```ts
$top=20&$orderby=Name asc&$filter=((Finish gt 2019-08-12T00:00:00Z and Finish lt 2019-08-19T00:00:00Z) and (PercentComplete ge 10 and PercentComplete le 100) and Gender eq 'male')
```
![image](https://user-images.githubusercontent.com/643976/62868324-5f101b80-bce3-11e9-9753-5bf5737c7dc7.png)

**Comment by hgc10120** [comment]
Unfortunately, I have no way of verifying whether this query works or not :(. The database that I'm querying on has no date fields.

**Comment by ghiscoding** [comment]
Thanks anyway, I'll assume it's ok until someone says it's not lol... but so far from all the docs I've seen, it seems to be it. Let's hope these 2 are the only ones that are different. 

I should be able to release a new version this week 🤞 , I'm at the final stage of the PR #252

**Comment by hgc10120** [comment]
Perfect. Also, huge thanks for putting Slick grid together. I've been exploring it a bit and it's been awesome so far.

**Comment by ghiscoding** [comment]
Yeah I used SlickGrid in the past, then we switched to UI-Grid in AngularJS which never got ported (neither will it, since it would require a complete rewrite). UI-Grid was really similar to SlickGrid (they probably copied their way of defining grid options and column definitions from SlickGrid), so at the end I came up with a mix of approach from UI-Grid and SlickGrid and now we use it in our company on a few projects. 

All I ask is maybe to up vote ⭐️ if you haven't already and spread the word 😉

**Comment by ghiscoding** [comment]
The fix is now pushed with the new version [2.10.1](https://github.com/ghiscoding/Angular-Slickgrid/releases).

**Comment by hgc10120** [comment]
Just checked it out and it's working beautifully so far. Thank you :).
